python/ray/serve/pipeline/deployment_node.py

```diff
@@ -169,7 +169,19 @@ def get_import_path(self):
     def to_json(self, encoder_cls) -> Dict[str, Any]:
         json_dict = super().to_json_base(encoder_cls, DeploymentNode.__name__)
         json_dict["deployment_name"] = self.get_deployment_name()
-        json_dict["import_path"] = self.get_import_path()
+        import_path = self.get_import_path()
+
+        error_message = (
+            "Class used in DAG should not be in-line defined when exporting"
+            "import path for deployment. Please ensure it has fully "
+            "qualified name with valid __module__ and __qualname__ for "
+            "import path, with no __main__ or <locals>. \n"
+            f"Current import path: {import_path}"
+        )
+        assert "__main__" not in import_path, error_message
+        assert "<locals>" not in import_path, error_message
+
+        json_dict["import_path"] = import_path
```

Comment:
are you removing this given the decision to do it at app level?

also, does this detection cover the case where a function returns a class?

Reply:
I think the easiest thing is to pass `enforce` flag to each relevant class' to_json implementation, i think leaving it per class works better for localizing implementation, ex: ClassNode and FunctionNode uses `import_path` str while `PipelineInputNode` enforces `preprocessor_import_path`.

Then we only need JSON serialization enforcement on the `get_ingress_deployment` call which is not needed for local testing; And only enforce import path on `to_yaml()` so JSON serializable serve dag with inline classes can still be locally tested with HTTP.

 > also, does this detection cover the case where a function returns a class?

you mean a top-level function that returns a class as value ? Let me add a test for this.

Reply:
added a test with class factory and verified it fails json serde.